From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Zignixx <cobra1989@gmail.com>
Date: Wed, 10 Nov 2021 12:05:14 +0100
Subject: [PATCH] Sand dupe fix fix (enable sand dupe again)


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 996f09eaf7d1d6c5e27e2bb2630a6bcb18854172..54a5823f4c0305c6d91895f2903635c3e2840ef8 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -2740,7 +2740,10 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, n
             }
 
             this.processPortalCooldown();
-            this.tickEndPortal(); // Paper - make end portalling safe
+            // Sand dupe fix fix start
+            // This should be not here to get the sand dupe work again?
+            // this.tickEndPortal(); // Paper - make end portalling safe
+            // Sand dupe fix fix end
         }
     }
 
diff --git a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
index 8336ea928faa92c6f58f8cdfb9faf1d8e26c9ccf..0772a080e599746c21aa05dd0b828bc96b2e01bb 100644
--- a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -106,11 +106,13 @@ public class FallingBlockEntity extends Entity {
 
     @Override
     public void tick() {
+        // Sand dupe fix fix start
         // Paper start - fix sand duping
-        if (this.isRemoved()) {
-            return;
-        }
+        // if (this.isRemoved()) {
+            // return;
+        // }
         // Paper end - fix sand duping
+        // Sand dupe fix fix end
         if (this.blockState.isAir()) {
             this.discard();
         } else {
@@ -143,11 +145,13 @@ public class FallingBlockEntity extends Entity {
 
             this.move(MoverType.SELF, this.getDeltaMovement());
 
+			// Sand dupe fix fix start
             // Paper start - fix sand duping
-            if (this.isRemoved()) {
-                return;
-            }
+            // if (this.isRemoved()) {
+                // return;
+            // }
             // Paper end - fix sand duping
+            // Sand dupe fix fix end
 
             // Paper start - Configurable EntityFallingBlock height nerf
             if (this.level.paperConfig.fallingBlockHeightNerf != 0 && this.getY() > this.level.paperConfig.fallingBlockHeightNerf) {
diff --git a/src/main/java/net/minecraft/world/level/block/EndPortalBlock.java b/src/main/java/net/minecraft/world/level/block/EndPortalBlock.java
index 62c2f947a77570228dfdf4dae16c64eb97ee2f40..345259e9950fa2b36e657272845d1504a65a920d 100644
--- a/src/main/java/net/minecraft/world/level/block/EndPortalBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/EndPortalBlock.java
@@ -52,11 +52,23 @@ public class EndPortalBlock extends BaseEntityBlock {
             if (worldserver == null) {
                 // return; // CraftBukkit - always fire event in case plugins wish to change it
             }
+			// Sand dupe fix fix start
+            // CraftBukkit start - Entity in portal
+            EntityPortalEnterEvent event = new EntityPortalEnterEvent(entity.getBukkitEntity(), new org.bukkit.Location(world.getWorld(), pos.getX(), pos.getY(), pos.getZ()));
+            world.getCraftServer().getPluginManager().callEvent(event);
+
+            if (entity instanceof ServerPlayer) {
+                ((ServerPlayer) entity).changeDimension(worldserver, PlayerTeleportEvent.TeleportCause.END_PORTAL);
+                return;
+            }
+            // CraftBukkit end
+            entity.changeDimension(worldserver);
 
             // Paper start - move all of this logic into portal tick
-            entity.portalWorld = ((ServerLevel)world);
-            entity.portalBlock = pos.immutable();
+            // entity.portalWorld = ((ServerLevel)world);
+            // entity.portalBlock = pos.immutable();
             // Paper end - move all of this logic into portal tick
+            // Sand dupe fix fix end
         }
 
     }
